performance tuning: Key metrics on performance overview (#8788)
pingcap/docs: grafana-performance-overview-dashboard.md

---
title: Key Metrics on Performance Overview

> summary: Learn key metrics displayed on the Grafana Overview dashboard.

Reviewer: ```suggestion
summary: Learn key metrics displayed on the Performance Overview dashboard.
```

---

> # key Metrics on Performance Overview

Reviewer: ```suggestion
# Key Metrics on Performance Overview
```

If you use TiUP to deploy the TiDB cluster, the monitoring system (Prometheus & Grafana) is deployed at the same time. For more information, see [TiDB Monitoring Framework Overview](/tidb-monitoring-framework.md).

The Grafana dashboard is divided into a series of sub dashboards which include PD, TiDB, TiKV, Node_exporter, Overview, Performance Overview, and so on. A lot of metrics are there to help you diagnose.

The Performance Overview dashboard orchestrates the metrics of TiDB, PD, and TiKV, and presents each of them in the following sections:

> - Overview: Database time and SQL execution time summary. Tuning by color, you can quickly identify the database load profile and the performance bottleneck.

Reviewer: ```suggestion
- Overview: Database time and SQL execution time summary. By checking different colors in the overview, you can quickly identify the database load profile and the performance bottleneck.
```

- Load profile: Key metrics and resource usage, including database QPS, connection information, the MySQL command types the application interactes with TiDB, database internal TSO and KV request OPS, and resource usage of the TiKV and TiDB.

- Top-down latency breakdown: Query latency versus connection idel time ratio, query latency breakdown, TSO/KV request latency during execution, breakdown of write latency within TiKV.

> With the Performance Overview Dashboard, you can analyze performance efficiently, and confirm whether the bottleneck of user response time is in the database. If the bottleneck is in the database, you can identify the bottleneck inside the database, with database time overview, load profile and SQL latency breakdown. For details, see [TiDB Performance Analysis and Tuning](/tidb-performance-analysis-and-tuning.md).

Author: ```suggestion
With the Performance Overview Dashboard, you can analyze performance efficiently, and confirm whether the bottleneck of user response time is in the database. If the bottleneck is in the database, you can identify the bottleneck inside the database, with database time overview, load profile and SQL latency breakdown.
```

The following sections illustrate the metrics on the Performance Overview dashboard.

## Database Time by SQL Type

- database time: Total database time per second
- sql_type: Database time consumed by each type of SQL statements per second

## Database Time by SQL Phase

- database time: Total database time per second

> - get token/parse/compile/execute: Database time consumed in four SQL processing phaseses

Reviewer: ```suggestion
- get token/parse/compile/execute: Database time consumed in four SQL processing phases
```

The SQL execution phase is in green and other phases are in red on general. If non-green areas are large, it means much database time is consumed in other phases than the execution phase and further cause analysis is required.

## SQL Execute Time Overview

- execute time: Database time consumed during SQL execution per second
- tso_wait: Concurrent TSO waiting time per second during SQL execution
- kv request type: Time waiting for each KV request type per second during SQL execution. The total KV request wait time might exceed SQL execution time, because KV requests are concurrent.

> Green metrics stand for common KV write requests (such as prewrite and commit), blue metrics stand for common read requests, and metrics in other colors stand for unexpected situations which you need to pay attention. For example, pessimistic lock KV requests are marked red and TSO waiting is marked dark brown. 

Reviewer: ```suggestion
Green metrics stand for common KV write requests (such as prewrite and commit), blue metrics stand for common read requests, and metrics in other colors stand for unexpected situations which you need to pay attention to. For example, pessimistic lock KV requests are marked red and TSO waiting is marked dark brown. 
```

> If non-blue or non-green areas are large, it means there is bottleneck during SQL execution. For example:

Reviewer: ```suggestion
If non-blue or non-green areas are large, it means there is a bottleneck during SQL execution. For example:
```

    - The duration from receiving a request from the client to TiDB till TiDB exeucting the request and returning the result to the client. In general, client requests are sent in the form of SQL statements; however, this duration can include the execution time of commands such as `COM_PING`, `COM_SLEEP`, `COM_STMT_FETCH`, and `COM_SEND_LONG_DATA`.
    - TiDB supports Multi-Query, which means the client can send multiple SQL statements at one time, such as `select 1; select 1; select 1;`. In this case, the total execution time of this query includes the execution time of all SQL statements.

- avg: Average time to execute all requests
- 99: P99 duration to execute all requests
- avg by type: Average time to execute all requests in all TiDB instances, collected by type: `SELECT`, `INSERT`, and `UPDATE`

## Connection Idle Duration

Connection Idle Duration indicates the duration of a connection being idle.

- avg-in-txn: Average connection idle duration when the connection is within a transaction
- avg-not-in-txn: Average connection idle duration when the connection is not within a transaction
- 99-in-txn: P99 connection idle duration when the connection is within a transaction

> - 99-not-in-txn: P99 connection idle duration when the connection is within a transaction

Reviewer: ```suggestion
- 99-not-in-txn: P99 connection idle duration when the connection is not within a transaction
```
